README.md
# Git2PDF

## Description

Git2PDF is a simple parser which can make github nested markdown file all in one PDF.

## Notice

Parse a markdown file which has a lot of url links can lead a very long time to make a PDF. The reason is Git2PDF need check every url status, if one links is broken which can throw a exception by **wkhtmltopdf** and it will faliure. One another is it need sleep 500ms every url check for github anti-crawler strategy.

## Installation

before you can use this lib, it need install **wkhtmltopdf** to parse html to pdf, install guid can found its [website](https://wkhtmltopdf.org/).

If [available in Hex](https://hex.pm/docs/publish), the package can be installed
by adding `git2pdf` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:git2pdf, "~> 0.3.0"}
  ]
end
```

## How to use

### Input

```iex> GitHubParse.markdown_to_pdf "https://github.com/remoteintech/remote-jobs/blob/master/README.md", 2, false```

```iex> GitHubParse.markdown_to_pdf "https://github.com/remoteintech/remote-jobs/blob/master/README.md", 2, false, ["--encoding", "utf-8", "-d", "300", "--zoom", "1.5", "--user-style-sheet", "/path/to/your/custom.css"]```

Git2PDF use wkhtmltopdf to generate pdf, the last param is wkhtmltopdf's params, your can find more in this [link](https://wkhtmltopdf.org/usage/wkhtmltopdf.txt)

### Output

```{"/var/folders/k7/w1zkb4cn0472dk8t_j699pyc0000gn/T/KXiE8hJZ.html", "/var/folders/k7/w1zkb4cn0472dk8t_j699pyc0000gn/T/KXiE8hJZ.pdf"}```

You can see this project README.md parse result in this [Git2PDF](git2pdf.pdf), and you can see it append the [CHANGELOG.md](CHANGELOG.md) to the [README.md](README.md).

The first is url, sencond is deep_level which mean how deepth which you want it get, the last is whether it check every url in markdown, mostly it doesn't need to check and you don't want to wait too long time.

[Change Log](CHANGELOG.md)


CHANGELOG.md
## Change Log
### V0.3.0
Add wkhtmltopdf params
### V0.2.0
Add deep_level(control markdown nested deepth) and is_check_url(control whether to check every urls in html for 404 errors which can lead parse failure) parameters


test/git2pdf_test.exs
defmodule Git2pdfTest do
  use ExUnit.Case
  doctest Git2pdf

  test "greets the world" do
    assert Git2pdf.hello() == :world
  end
end


mix.exs
defmodule Git2pdf.MixProject do
  use Mix.Project

  def project do
    [
      app: :git2pdf,
      version: "0.3.0",
      elixir: "~> 1.7",
      build_embedded: Mix.env == :prod,
      start_permanent: Mix.env() == :prod,
      description: description(),
      package: package(),
      deps: deps(),
      name: "Git2PDF",
      source_url: "https://github.com/madawei2699/git2pdf"
    ]
  end

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:logger]
    ]
  end

  defp description() do
    "print github nested markdown file all in one pdf"
  end

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      {:pdf_generator, ">=0.4.0" },
      {:floki, "~> 0.20.0"},
      {:httpoison, "~> 1.4"},
      {:ex_doc, ">= 0.0.0", only: :dev},
      {:credo, "~> 1.0.0", only: [:dev, :test], runtime: false}
      # {:dep_from_hexpm, "~> 0.3.0"},
      # {:dep_from_git, git: "https://github.com/elixir-lang/my_dep.git", tag: "0.1.0"},
    ]
  end

  defp package() do
    [
      # This option is only needed when you don't want to use the OTP application name
      name: "git2pdf",
      # These are the default files included in the package
      files: ~w(lib config .formatter.exs mix.exs README* LICENSE* CHANGELOG*),
      licenses: ["MIT License"],
      links: %{"GitHub" => "https://github.com/madawei2699/git2pdf"}
    ]
  end
end


config/config.exs
# This file is responsible for configuring your application
# and its dependencies with the aid of the Mix.Config module.
use Mix.Config

# This configuration is loaded before any dependency and is restricted
# to this project. If another project depends on this project, this
# file won't be loaded nor affect the parent project. For this reason,
# if you want to provide default values for your application for
# 3rd-party users, it should be done in your "mix.exs" file.

# You can configure your application as:
#
#     config :git2pdf, key: :value
#
# and access this configuration in your application as:
#
#     Application.get_env(:git2pdf, :key)
#
# You can also configure a 3rd-party app:
#
#     config :logger, level: :info
#

# It is also possible to import configuration files, relative to this
# directory. For example, you can emulate configuration per environment
# by uncommenting the line below and defining dev.exs, test.exs and such.
# Configuration from the imported file will override the ones defined
# here (which is why it is important to import them last).
#
#     import_config "#{Mix.env()}.exs"


lib/github_parse.ex
defmodule GitHubParse do

  @moduledoc """
  Git2PDF is a simple parser which can make github nested markdown file all in one PDF.
  """

  @doc """
  ## Example
    iex> GitHubParse.markdown_to_pdf "https://github.com/remoteintech/remote-jobs/blob/master/README.md", 2, false
    iex> GitHubParse.markdown_to_pdf "https://github.com/remoteintech/remote-jobs/blob/master/README.md", 2, false, ["--encoding", "utf-8", "-d", "300", "--zoom", "1.5", "--user-style-sheet", "/path/to/your/custom.css"]

  """
  def markdown_to_pdf(url, deep_level, is_check_url, wkhtmltopdf_options \\ ["--encoding", "utf-8"]) do
    readme_html = url_to_html(url, false, deep_level, is_check_url)
    pdf_file_path = PdfGenerator.generate! readme_html, shell_params: wkhtmltopdf_options
    html_file_path = Enum.at(String.split(pdf_file_path, "."), 0) <> ".html"
    File.write!(html_file_path, readme_html)
    {html_file_path, pdf_file_path}
  end

  defp get_html_body(url) do
    case HTTPoison.get(url) do
      {:ok, %HTTPoison.Response{status_code: 200, body: body}} ->
        body
      _ ->
        ""
    end
  end

  defp url_to_html(url, is_readme, deep_level, is_check_url, url_list \\ [])

  defp url_to_html(url, is_readme, deep_level, is_check_url, url_list) when deep_level > 0 do
    IO.puts("url is #{url}")
    # must include else clause or it will nil
    url_list = if url_list == [] do List.insert_at(url_list, 0, url) else url_list end
    Process.sleep(Enum.random(1..2) * 500)
    body = get_html_body(url)
    readme =
      if is_readme do
        Enum.at(Floki.find(body, "#readme div"), 1)
      else
        Floki.find(body, "#readme")
      end
    normal_html =
      if is_readme do
        walk_html_node(readme, is_check_url)
      else
        walk_html(readme, is_check_url)
      end
    new_url_list = get_url_list(Floki.find(normal_html, "a"))
    IO.puts("new url list is #{new_url_list}")
    child_htmls = new_url_list
      |> Stream.filter(fn x -> not Enum.member?(url_list, x) end)
      |> Stream.map(&url_to_html(&1, false, deep_level - 1, is_check_url, Enum.concat(new_url_list, url_list)))
      |> Enum.to_list
    [Floki.raw_html(flatten_list(normal_html)) | child_htmls]
  end

  defp url_to_html(_, _, deep_level, _, _) when deep_level == 0 do
    []
  end

  defp get_url_list(html) do
    html
      |> Floki.attribute("href")
      |> Enum.filter(fn x -> String.contains?(x, ".md") and not String.contains?(x, ".md#") end)
      |> Enum.uniq
  end

  defp flatten_list(list) do
    case list do
      [[l]] -> [l]
      other -> other
    end
  end

  defp walk_html(html, is_check_url) do
    html
    |>Enum.map(&walk_html_node(&1, is_check_url))
    |>Enum.to_list
  end

  defp walk_html_node(node, is_check_url) do
    case node do
      {tag, attribute_list, child_node_list} -> {tag, walk_html_attr(attribute_list, is_check_url), walk_html_node(child_node_list, is_check_url)}
      list when is_list(list) -> walk_html(list, is_check_url)
      _ -> node
    end
  end

  defp walk_html_attr(attr_list, is_check_url) do
    attr_list
    |>Enum.map(fn(item) -> fix_url(item, is_check_url) end)
    |>Enum.to_list
  end

  defp fix_url({name, value}, is_check_url) do
    case name do
      "src" -> if not String.starts_with?(value, "http") and not String.starts_with?(value, "#") do {name,  check_url_status("https://github.com" <> value, is_check_url)} else {name, check_url_status(value, is_check_url)} end
      "href" -> if not String.starts_with?(value, "http") and not String.starts_with?(value, "#") do {name, check_url_status("https://github.com" <> value, is_check_url)} else {name, check_url_status(value, is_check_url)} end
      _ -> {name, value}
    end
  end

  defp check_url_status(url, is_check_url) when is_check_url == false do
    url
  end

  defp check_url_status(url, is_check_url) when is_check_url == true do
    if String.starts_with?(url, "http") do
      IO.puts("check_url_status: url is #{url}")
      Process.sleep(500)
      r = HTTPoison.head url
      case r do
        {:ok, %HTTPoison.Response{status_code: 404, body: _}} ->
          IO.puts "url is null"
        _ ->
          url
      end
    else
      ""
    end
  end
end
